### PR TITLE
end-to-end  gpu driver testing enhancement

### PR DESCRIPTION
**Pull request by shivakunv**
TODO--

*1. *
> + matrix:
+ driver:
+ - 535.183.06
+ - 550.90.07
An idea for a potential follow-up – instead of defining a matrix and spinning up one AWS instance per driver version, can we instead pass all driver versions as input to the test script and test all of them in sequence? For example, first install 535.183.06, and then upgrade to 550.90.07? This would also allow us to source the driver versions from the versions.mk file, instead of having to redefine and maintain that list here.

2
+name: CI
Let's rename this to End-to-end tests

3
+# Install the operator with usePrecompiled mode set to true
Remove this comment as it is not accurate.

4
+ AWS_SESSION_TOKEN: ${{ secrets.AWS_SESSION_TOKEN }}
I don't believe AWS_SESSION_TOKEN is needed with current holodeck implementation. Let's remove.

5
We can simply wait for the nvidia-driver pod to be ready
kubectl wait -n ${TEST_NAMESPACE} --for=condition=Ready pod -l app=nvidia-driver-daemonset --timeout 10m
If successful, then wait for the validator pod to be ready (this means that the rest of the pods are healthy):
kubectl wait -n ${TEST_NAMESPACE} --for=condition=Ready pod -l app=nvidia-operator-validator --timeout 2m
If either of these commands fails, capture the state of all pods in the operator namespace, by running kubectl get pods -n ${TEST_NAMESPACE} , and also capture some logs so we can debug.
This will reduce the amount of logs emitted during the test. Right now, we print out all pods every 5 seconds so it is very unreadable

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -22,14 +22,14 @@ on:
     branches:
       - main
 
+  # SHIVA
```
Comment (shivakunv) [inline .github/workflows/ci.yaml:25]: will remove for testing purpose only

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -59,11 +58,34 @@ jobs:
...
         TEST_CASE: "./tests/cases/nvidia-driver.sh"
       run: |
         sudo chmod 644 ${{ github.workspace }}/.cache/key
         echo "${{ secrets.AWS_SSH_KEY }}" > ${private_key} && chmod 400 ${private_key}
-        ./tests/ci-run-e2e.sh ${TEST_CASE} ${COMMIT_SHORT_SHA}-${{ matrix.driver }}
+        rc=0
+        for driver_version in ${{ env.DRIVER_VERSIONS }}; do
+          echo "Running e2e for DRIVER_VERSION=$driver_version"
+          # ./tests/ci-run-e2e.sh ${TEST_CASE} ${COMMIT_SHORT_SHA}-${driver_version}
+          # SHIVA
```
Comment (shivakunv) [inline .github/workflows/ci.yaml:73]: will remove , for testing purpose

File: .github/workflows/image.yaml [hunk 3]
```diff
@@ -93,86 +93,89 @@ jobs:
           IMAGE_NAME: ghcr.io/nvidia/driver
           VERSION: ${COMMIT_SHORT_SHA}
         run: |
-          DRIVER_VERSIONS=${{ matrix.driver }} make build-${{ matrix.dist }}-${{ matrix.driver }}
+          # SHIVA
+          # DRIVER_VERSIONS=${{ matrix.driver }} make build-${{ matrix.dist }}-${{ matrix.driver }}
```
Comment (shivakunv) [inline .github/workflows/image.yaml:97]: for testing purpose

File: tests/scripts/verify-operator.sh [hunk 4]
```diff
@@ -11,9 +11,39 @@ source ${SCRIPT_DIR}/.definitions.sh
...
+# collect log in case of failure
+local start_time=$(date +%s)
+local log_dir=${LOG_DIR}
+while :; do
+    current_time=$(date +%s)
+    elapsed_time=$((current_time - start_time))
+    
+    # Check if total elapsed time is greater than exit
+    if [ $elapsed_time -gt $MAX_POD_STATUS_CHECK_TOTAL_TIME ]; then
+        echo "Total wait time exceeded ${MAX_POD_STATUS_CHECK_TOTAL_TIME} seconds. Exiting..."
```
Comment (shivakunv) [inline tests/scripts/verify-operator.sh:25]: exit when 50 minutes completed

File: tests/scripts/checks.sh [hunk 5]
```diff
@@ -2,35 +2,32 @@
...
-		if [[ "${current_time}" -gt $((60 * 45)) ]]; then
-			echo "timeout reached"
-			exit 1;
-		fi
+	return 1
+}
 
-		# Echo useful information on stdout
-		kubectl get pods -n ${TEST_NAMESPACE}
+cleanup_pod(){
```
Comment (cdesiniotis) [inline tests/scripts/checks.sh:23]: Instead of implementing this function, create a new script file named `uninstall-operator.sh` that uninstalls the operator. Uninstalling the chart will remove all of the resources in the namespace, so you just need to run:

```
helm uninstall $release_name --namespace $namespace
```

File: tests/scripts/remote.sh [hunk 6]
```diff
@@ -4,4 +4,5 @@ SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 source ${SCRIPT_DIR}/.definitions.sh
 source ${SCRIPT_DIR}/.local.sh
 
-ssh -i ${private_key} ${instance_hostname} "${@}"
+# keep alive 60sec and timeout after 30 tries
+ssh -o ServerAliveInterval=60 -o ServerAliveCountMax=30 -i ${private_key} ${instance_hostname} "${@}"
```
Comment (cdesiniotis) [inline tests/scripts/remote.sh:8]: Question -- why is this required?

File: tests/scripts/verify-operator.sh (hunk 4 above)
Comment (cdesiniotis) [reply tests/scripts/verify-operator.sh:26]: Why do we need a while loop and all of this extra logic? Running `kubectl --wait` for both the driver and validator pod should be all we need, and we can exit when a timeout is reached.

Can we simplify this entire file to just:
```
kubectl --wait nvidia-driver-daemonset ...
kubectl --wait nvidia-operator-validator ...

if rc != 0 {
    collect_logs
}
```

And we can extract the `collect_logs` function to its own script.

File: tests/scripts/verify-operator.sh [hunk 7]
```diff
@@ -11,9 +11,42 @@ source ${SCRIPT_DIR}/.definitions.sh
...
+    fi
+
+    check_pod_ready "nvidia-driver-daemonset" ${DAEMON_POD_STATUS_TIME_OUT} && \
+        check_pod_ready "nvidia-operator-validator" ${POD_STATUS_TIME_OUT}; status=$?
+    
+    if [ $status -ne 0 ]; then
+        not_ready_pod_status=$(kubectl get pods -n $TEST_NAMESPACE -o jsonpath="{.items[?(@.status.phase != 'Running')].metadata.name}")
+        if [ -n "$not_ready_pod_status" ]; then
+            for pod in $not_ready_pod_status; do
+                echo "Collecting logs for pod: $pod in dir ${log_dir}"
```
Comment (cdesiniotis) [inline tests/scripts/verify-operator.sh:38]: As mentioned in a previous comment, lets extract this into a separate function, `collect_logs`, which we can define in a separate file.

Note, we have a script named `must-gather.sh` which we ask users to run when they are hitting operator issues. It collects a comprehensive set of logs to help us debug. Feel free to reuse some of the logic from there, specifically the logic to collect pod logs: https://github.com/NVIDIA/gpu-operator/blob/main/hack/must-gather.sh#L169-L196. If it makes sense, we can consider just running the `must-gather.sh` script directly from within our tests to collect logs.

Comment (cdesiniotis) [reply tests/scripts/verify-operator.sh:38]: Also, I downloaded the logs generated from this failing pipeline, and all of the files are empty: https://github.com/shivakunv/gpu-driver-container/actions/runs/10425570121
Can you check this?

File: tests/scripts/checks.sh (hunk 5 above)
Comment (shivakunv) [reply tests/scripts/checks.sh]: uninstall-operator.sh  created . 
helm uninstall $release_name --namespace $namespace was already there at the end of the function . 
_Uninstalling Helm itself will not remove DaemonSets or Secrets_

File: tests/scripts/remote.sh (hunk 6 above)
Comment (shivakunv) [reply tests/scripts/remote.sh:8]: default ssh server timeout 60sec
_kubectl wait or sleep  :
 ssh might drop the connection if there's no activity between the client and server ( default timeout 60sec)_

File: tests/scripts/verify-operator.sh (hunk 7 above)
Comment (shivakunv) [reply tests/scripts/verify-operator.sh]: I removed the while loop and added more delay since the gpu-operator sometimes takes longer to deploy. 
increased the daemon wait time to 15 minutes.

I kept the must-gather.sh script as it is, as it only requires the artifacts directory path for log collection. 
I have integrated it with the current code

**Comment by shivakunv** [comment]
@cdesiniotis PTAL

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -21,15 +21,10 @@ on:
       - completed
     branches:
       - main
-
+
```
Comment (cdesiniotis) [inline .github/workflows/ci.yaml:24]: nit: remove this unneeded change.

File: tests/scripts/must-gather.sh [hunk 9]
```diff
@@ -0,0 +1,256 @@
+#!/usr/bin/env bash
```
Comment (cdesiniotis) [inline tests/scripts/must-gather.sh:1]: Instead of maintaining a copy of this script here, let's download this script on the remote machine at runtime.

File: tests/scripts/uninstall-operator.sh [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+source ${SCRIPT_DIR}/.definitions.sh
+
+namespace=$1
+release_name=$2
+kubectl delete daemonsets --all -n $namespace || true
+kubectl delete deployments --all -n $namespace || true
+kubectl delete services --all -n $namespace || true
+kubectl delete pods --all -n $namespace || true
+kubectl delete namespace $namespace || true
```
Comment (cdesiniotis) [inline tests/scripts/uninstall-operator.sh:17]: These `kubectl delete` commands are not needed, as the helm uninstall should delete all resources in the namespace. If desired, you can delete the namespace after uninstalling the helm chart.

File: tests/scripts/must-gather.sh (hunk 9 above)
Comment (shivakunv) [reply tests/scripts/must-gather.sh]: added

**Review by cdesiniotis — APPROVED** [review]
